METRON-1038: Stellar should have a better collection of basic math operations

cestella opened the pull request:
## Contributor Comments
At the moment the math functions are woefully incomplete. 
We should add at least the ones difficult or impossible to reconstruct using existing stellar primitives/math functions:
* log10
* log2
* ln
* sqrt
* ceil
* floor
* sin
* cos
* tan

It should be sufficient to test these out in the Stellar REPL for manual evaluation.  I have also refactored and added a more robust suite of unit tests for existing basic math functions as well as these new ones.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

ottobackwards commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java:
```diff
@@ -25,17 +25,39 @@
-          , returns="The absolute value of the number passed in."
-          )
-  public static class Abs implements StellarFunction {
+  private enum SingleArgMathFunctions implements Function<Double, Double> {
```
Why have this private? why not have stellar/common/util/math/*

cestella replied:
I can certainly pull this out if you think it's useful.  I didn't think it was super useful outside of a helper enum for this class.

ottobackwards commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java:
```diff
@@ -25,17 +25,39 @@
+    SQRT(d -> Math.sqrt(d)),
+    TAN(d -> Math.tan(d)),
+    ;
+
```
Is there a reason why only this subset of Math.* is exposed?  What about the rest?  Are they not needed, applicable, or left for future work?

cestella replied:
Honestly, I just added the ones that seem to be the most pressingly missing and that we could not recreate using primitives.  Did you have any functions that you wanted in there?

cestella replied:
I think they're all applicable, most of the remainder aren't as useful to Metron, IMO, but I would be willing to go through and add them if we feel that there is a need for completeness.

ottobackwards replied:
I agree.  I think that in general when we are adapting or exposing a set of functionality  from java or from apache-commons  we should state our intent explicitly for the record.

cestella commented:
Ok, I broke this abstraction out a bit and made it easier to add non-single-arg functions (a la POW, which is a separate PR by @simonellistonball ).  I also went ahead and added `ROUND` and `EXP`.  I looked through the remainder of the math functions and honestly, I didn't see a strong reason to include any of the rest of them immediately.  I'd RATHER them come in via separate PRs with some justification of use.

cestella commented:
Any comments or review here?

simonellistonball commented:
I would like it to be committed to make it much easier for me to port my PR to it.... so +1 (non-binding)

justinleet commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java:
```diff
@@ -19,45 +19,179 @@
+  }
+
+  @Stellar(name="ROUND"
+          ,description="Rounds a number to the nearest integer"
```
Would you mind mentioning explicitly that this is half-up rounding?  That's going to be the default assumption, but it should be explicit since we aren't supporting other modes for right now.

mattf-horton replied:
I'm okay not documenting this, since it is the norm, but if documented here do it in the .md file too.

cestella replied:
Yeah, I documented it in the most recent commit.

mattf-horton commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/math/StellarMathFunction.java:
```diff
@@ -0,0 +1,64 @@
+
+  @Override
+  public Object apply(List<Object> args, Context context) throws ParseException {
+    if(args.size() < _func.getMaxArgs()) {
```
If this is the logical use of getMaxArgs() then it should be called minRequiredArgs(), shouldn't it?  Or perhaps you need another member for that.

cestella replied:
Agreed, we need `getMinArgs()`.  This is confusing; good catch.

mattf-horton commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/math/MathOperation.java:
```diff
@@ -0,0 +1,39 @@
+  public MathOperation(Function<Number[], Number> operation, int maxArgs) {
+    this.operation = operation;
+    this.maxArgs = maxArgs;
+  }
```
If you give equal billing to binary ops in MathOperations.java, as requested below, then this also needs an additional constructor with BiFunction arg, I think.

mattf-horton commented on metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/MathFunctionsTest.java:
```diff
@@ -38,10 +46,156 @@ public static Object run(String rule, Map<String, Object> variables) {
+    for(int i = 1;i <= 10;++i) {
+      expectedValues.put(Math.pow(base, i), (double)i);
+    }
+    assertValues(logExpr, expectedValues);
```
These test look pretty good.

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -110,6 +110,8 @@ In the core language functions, we support basic functional programming primitiv
 | [ `BLOOM_EXISTS`](#bloom_exists)                                                                   |
 | [ `BLOOM_INIT`](#bloom_init)                                                                       |
 | [ `BLOOM_MERGE`](#bloom_merge)                                                                     |
+| [ `CEIL`](#ceil)                                                                                   |
```
Yep, changing it.

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -110,6 +110,8 @@ In the core language functions, we support basic functional programming primitiv
 | [ `BLOOM_INIT`](#bloom_init)                                                                       |
 | [ `BLOOM_MERGE`](#bloom_merge)                                                                     |
+| [ `CEIL`](#ceil)                                                                                   |
+| [ `COS`](#ceil)                                                                                   |
```
Thanks, good catch.

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -238,6 +249,18 @@ In the core language functions, we support basic functional programming primitiv
+  * Returns: The ceiling of the number passed in.
+
+### `COS`
+  * Description: Returns the `cos` of a number.
```
yep

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -581,6 +634,24 @@ In the core language functions, we support basic functional programming primitiv
+  * Returns: The nearest integer.
+
+### `SIN`
+  * Description: Returns the `sin` of a number.
```
yep

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -581,6 +634,24 @@ In the core language functions, we support basic functional programming primitiv
+  * Returns: The `sin` of the number passed in.
+
+### `SQRT`
+  * Description: Returns the `sqrt` of a number.
```
yep

cestella commented on metron-stellar/stellar-common/README.md:
```diff
@@ -613,6 +684,12 @@ In the core language functions, we support basic functional programming primitiv
   * Returns: String
 
+### `TAN`
+  * Description: Returns the `tan` of a number.
```
yep :)

cestella commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/math/MathOperations.java:
```diff
@@ -0,0 +1,66 @@
+  MathOperation op;
+  MathOperations(Function<Double, Number> singleArg) {
+    op = new MathOperation(new SingleOpFunc(singleArg), 1);
+  }
```
Yeah, I'll add binary function too

cestella commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/math/MathOperations.java:
```diff
@@ -0,0 +1,66 @@
+    public Number apply(Number[] numbers) {
+      return f.apply(numbers[0].doubleValue());
+    }
+  }
```
yep, I'll add it here; makes sense.

cestella commented:
Ok, I reacted to @mattf-horton  and @justinleet 's comments:
* adjusted some naming and made the documentation match up
* adjusted the `ROUND` documentation to indicate half-up rounding.
* added some abstraction candy around binary functions

Let me know what you guys think.

jjmeyer0 commented on metron-stellar/stellar-common/README.md:
```diff
@@ -101,58 +101,69 @@ In the core language functions, we support basic functional programming primitiv
+| [ `KAFKA_PUT`](#kafka_put)                                                       |
+| [ `KAFKA_TAIL`](#kafka_tail)                                                     |
+| [ `LENGTH`](#length)                                                             |
+<<<<<<< HEAD
```
@cestella looks like something may have happened with a merge.

cestella replied:
yep, fixed.  Good eye :)

mattf-horton commented:
+1, lgtm.  By the way, my mentions of BiFunction were incorrect because MathFunction takes an array of numbers as its (sole) argument.  But glad to see you added Min Args as well as Max Args.

cestella commented:
Committed, but I wanted to make a comment about the BiFunction.  Honestly, I like the notion of providing candy for 2-arg functions like we did for single-arg functions.  I think it'll make it super easy for @simonellistonball to add `POWER`.
